CombinatorialGames/Surreal/Numeric.lean

```diff
@@ -226,6 +270,143 @@ theorem upper_neg (x : Dyadic) : upper (-x) = -lower x := by
   unfold lower upper
   simp [Rat.neg_mkRat, ← sub_eq_neg_add]
 
+theorem num_den_add_self_of_den_ne_one {x : Dyadic} (hx : x.1.den ≠ 1) :
+    (x + x).1.num = x.1.num ∧ (x + x).1.den = x.1.den / 2 := by
+  obtain ⟨r, hr⟩ := x
+  cases r using Rat.numDenCasesOn'' with | _ n d nz red =>
+  dsimp [IsDyadic] at hr hx ⊢
+  rw [Submonoid.mem_powers_iff] at hr
+  obtain ⟨⟨⟩ | c, rfl⟩ := hr
+  · simp at hx
+  · have hc₀ : 2 ^ (c + 1) / 2 ≠ 0 := by simp [Nat.pow_add_one]
+    have hcc : n.natAbs.Coprime (2 ^ (c + 1) / 2) := by
+      rw [Nat.coprime_pow_right_iff c.zero_lt_succ] at red
+      rw [Nat.pow_add_one, Nat.mul_div_cancel (2 ^ c) zero_lt_two]
+      exact red.pow_right c
+    set d : Rat := ⟨n, 2 ^ (c + 1), nz, red⟩
+    rw [← Rat.mk'.injEq _ _ (d + d).den_nz (d + d).reduced n _ hc₀ hcc]
+    simp_rw [Rat.mk'_num_den, d, Rat.mk'_eq_divInt, ← Rat.add_divInt]
+    rw [Nat.pow_add_one, Nat.mul_div_cancel (2 ^ c) zero_lt_two,
+      Rat.divInt_eq_divInt (by simp) (by simp)]
+    rw [Nat.cast_mul, Nat.cast_pow, Nat.cast_ofNat, ← mul_two, mul_right_comm, mul_assoc]
+
+theorem den_add_intCast {x : Dyadic} (n : ℤ) : (x + n).1.den = x.1.den := by
+  obtain ⟨x, hx⟩ := x
+  dsimp
+  cases x using Rat.numDenCasesOn' with | _ t b nz =>
+  rw [Rat.intCast_eq_divInt, Rat.divInt_add_divInt t n (Nat.cast_ne_zero.mpr nz) one_ne_zero,
+    mul_one, mul_one, Rat.den_mk, Rat.den_mk]
+  rw [if_neg (Nat.cast_ne_zero.mpr nz), if_neg (Nat.cast_ne_zero.mpr nz)]
+  simp_rw [Int.gcd_eq_natAbs]
+  apply congrArg (b / ·)
+  apply Nat.dvd_antisymm
+  · rw [Nat.dvd_gcd_iff]
+    constructor
+    · rw [← Int.gcd_eq_natAbs, ← Int.ofNat_dvd, Int.dvd_natAbs]
+      conv_rhs => equals t + n * b - n * b => simp
+      apply dvd_sub
+      · exact Int.gcd_dvd_left
+      · apply dvd_mul_of_dvd_right
+        exact Int.gcd_dvd_right
+    · apply Nat.gcd_dvd_right
+  · rw [Nat.dvd_gcd_iff]
+    constructor
+    · rw [← Int.gcd_eq_natAbs, ← Int.ofNat_dvd, Int.dvd_natAbs]
+      apply dvd_add
+      · exact Int.gcd_dvd_left
+      · apply dvd_mul_of_dvd_right
+        exact Int.gcd_dvd_right
+    · apply Nat.gcd_dvd_right
+
+theorem den_zsmul_le {x : Dyadic} (n : ℤ) : (n • x).1.den ≤ x.1.den := by
+  obtain ⟨x, hx⟩ := x
+  dsimp
+  cases x using Rat.numDenCasesOn' with | _ t b nz =>
+  rw [zsmul_eq_mul, Rat.intCast_eq_divInt, Rat.divInt_mul_divInt', one_mul,
+    Rat.den_mk, Rat.den_mk]
+  rw [if_neg (Nat.cast_ne_zero.mpr nz), if_neg (Nat.cast_ne_zero.mpr nz)]
+  apply Nat.div_le_div_left
+  · apply Nat.le_of_dvd (by simp [Int.gcd_pos_iff, nz])
+    exact Int.gcd_dvd_gcd_mul_left t b n
+  · simp [Int.gcd_pos_iff, nz]
+
+@[simp]
+theorem lower_add_self {x : Dyadic} (hx : x.1.den ≠ 1) :
```

Comment:
I think these lemmas on `lower` and `upper` might be nice to have. They might help us if we ever want to work with dyadic games explicitly in the future.